### DISCUSSION TITLE
Native i18n support in TinaCMS

### DISCUSSION THREAD

**Discussion by ccapndave** [tinacms/tinacms#4003] [Ideas]
I am thinking about porting a large PHP site over to NextJS and TinaCMS.  However we need to support different languages, and currently there is no smooth way to do this with TinaCMS.

I thought I would write down my thoughts about what would work for our use case, and then perhaps other people can chime in with what they would need and we can figure out a good way forward.

Our requirements boil down to this:

 - certain fields (but not all of them) should be translated
 - it should be possible to edit different languages directly in TinaCMS
 - translations should be stored in external files (i.e. *not* the core data files) so that they can be translated using software like https://crowdin.com/

At its most simple, I can see this working by adding something like an optional `translateKey` property to every field type in the schema.  If the property isn't set then things work as normal.  However, if the property is set to a key (e.g. `page.home.title`), then the CMS will read and write the value of the field from a nested `page.home.title` key in a `/messages/<locale>.json` file.

The job of actually reading and rendering the translated content is already solved by multiple libraries and frameworks, and I believe that any of them will be happy to read translations from JSON files containing nested keys so they could hopefully work out of the box.

Does this sound like something that could work?

**Comment by twittkopf** [comment]
I did it like this in the past:

- Translations for string in the UI (like button labels) are handled by a translation framework that consumes json files. These files can easily be synced with a translation software like https://crowdin.com
- The actual content of the page (like blog posts) is entered into Tina in the appropriate language.

This felt natural to me because in a way, I think this is how content management should work.

From your message I get the impression you want to use Tina like a translation software, is that correct?

Without testing any of this, I think you could make this work today. By setting a collection's format to `json` and using your translation keys as field names, you get something similar to a translation file. It may need some transforms to actually work as translation because depending on your content model, the file created by Tina probably also contains parts your translation framework doesn't understand.

**Reply by ccapndave** [reply]
Thanks for your response!

> From your message I get the impression you want to use Tina like a translation software, is that correct?

My perspective is that content is content, whatever language its in, and its absolutely valid to use a CMS to edit a site in different languages.  On a practical note, the translators also want to check their translations in Tina to confirm that changes in translation lengths doesn't break the design.

> By setting a collection's format to json and using your translation keys as field names, you get something similar to a translation file.

I've spent quite a lot of time trying this and Crowdin simply can't be configured to ignore nested `_template` keys.  A competing translation platform called Phrase can ignore them with a regexp, but its hacky and fragile.  It also means that the translation software could accidentally change the structure of a page, since that is bundled in with the translations.
